SQLite3 Adapter

AndrewJo opened the pull request:
## SQLite3 Adapter

First of all, I'd like to thank the Sails.js team for creating such an amazing framework. I ended up creating a SQLite3 adapter for Waterline because I needed a lightweight database that was going to run on an embedded device (BeagleBone Black) and it seemed like there was no support for it.

Again, special thanks to the developers and community of Sails.js and Node!
### Important Note

**Please do not merge this yet** as the code is definitely not ready for primetime. None of the unit tests have been written for it and lot of the code was loosely based on the PostgreSQL adapter which may or may not cause problems with SQLite. I am submitting the pull request to let everyone know that the SQLite adapter is under active development and to eliminate duplicate efforts (and possibly get some helping hand on my end).

AndrewJo committed:
Updated title and main description

AndrewJo committed:
Updated package name and authorship information

AndrewJo committed:
Added sqlite3 as dependency

AndrewJo committed:
Added underscore.js

AndrewJo committed:
Added disclaimer

AndrewJo committed:
Initial commit

AndrewJo committed:
Added reference to Query objects

AndrewJo committed:
Initial commit

AndrewJo committed:
Added buildSelectStatement function

AndrewJo committed:
Initial commit

AndrewJo committed:
Initial commit

AndrewJo committed:
Added unit test packages

AndrewJo committed:
Fixed namespace typo

AndrewJo committed:
Initial commit

AndrewJo committed:
Removed index.js

AndrewJo committed:
Fixed a typo in the regular expression and added a return statement

AndrewJo committed:
Prevent raising errors for dropping tables that don't exist

AndrewJo committed:
Switched to named parameters for index list query and fixed callback issue

AndrewJo committed:
Switched to named parameters for index list query and fixed callback issue

AndrewJo committed:
Initial commit

AndrewJo committed:
Initial unit test commit

AndrewJo committed:
Fixed bugs found from unit tests

AndrewJo committed:
Updated async version

AndrewJo committed:
Added sqlite3 module and typo fix to db config

AndrewJo committed:
Added new method and typo fix for closing db

AndrewJo committed:
Major bug fixes found from unit tests

AndrewJo committed:
Initial commit

AndrewJo committed:
Fixed find(), update() bugs

AndrewJo committed:
Added Seed() method

AndrewJo committed:
Fixed a bug in destroy() method

AndrewJo committed:
Fixed a bug with creating index

AndrewJo committed:
Indexing unit test

AndrewJo committed:
Aggregate unit tests

AndrewJo committed:
Added query unit tests

AndrewJo committed:
Added aggregate query unit tests

AndrewJo committed:
Fixed a spacing typo

olivierFr commented:
nice work! it would be really nice to be able to use Sqlite with Sails. On some of my projects, Sqlite is mandatory.

coderbuzz commented:
Any progress for this??? NO official SQLite adapter? Come on!

PostgreSQL - 0.9+ compatible
MySQL - 0.9+ compatible
MongoDB - 0.9+ compatible
Memory - 0.9+ compatible
Disk - 0.9+ compatible
Microsoft SQL Server
Redis
Riak
IRC
Twitter
JSDom
Neo4j
OrientDB
FoundationDB SQL Layer
ArangoDB
Apache Cassandra

dmarcelino commented:
Hi @coderbuzz,

The adapters officially supported by @balderdashy are currently: PostgreSQL, MySQL, MongoDB, Memory, Disk and Redis. I haven't heard of any plans to support more adapters, still all community adapter efforts are welcomed!

There was a similar discussion in balderdashy/waterline#405 and you may want to look at https://github.com/AndrewJo/sails-sqlite3. If you do create a fully functional sqlite adapter that passes the [API tests](https://github.com/balderdashy/waterline-adapter-tests) please let us know and we'll add it to the waterline README together with the adapters you mentioned.

tjwebb commented:
> Any progress for this??? NO official SQLite adapter? Come on!

This is a really terrible attitude. If you'd like to contribute then we'd be happy to collaborate with you, but complaining that other people on the internet haven't built what you want for free isn't going to inspire anyone to help you.

coderbuzz commented:
Hi,

Sorry my terrible English, I'm not going complain to the TS (@AndrewJo)

I know SailsJS year ago if not mistaken. Yesterday I'm going to search any new 2015 Node.JS's Web App Framework, then I landed here (again), found Waterline is now as separated module, but...

Every ORM frameworks usually come with SQLite adapter as default, and I don't see this in Waterline.
Couldn't understand why SQLite is not in your (Waterline) planning/roadmap, SQLite is used (almost) anywhere.

Forgive my astonishment (Google translate say this, astonishment -> is this correct? :D )

vedranstanic82 commented:
Hi all,

I was planing on giving Sails a go, but without this adapter i cannot proceed which is very unfortunate. Good luck with implementing it!

Vedran

vedranstanic82 commented:
I wonder if something could be reused from this project: https://github.com/Synerzip/loopback-connector-sqlite

Vedran

tjwebb commented:
I'm sorry that this is a blocker for some people. We rely on the community to extend sails to suit their own needs. There are dozens of database adapters, and there's no way Balderdash could build and maintain each one of those ourselves.

It seems like there's a lot of interest, and if everyone here collaborated on an adapter it could probably be implemented very quickly. This is one of the great strengths of open-source software. This is probably a decent starting point: https://github.com/AndrewJo/sails-sqlite3/tree/0.10.

There are also a number of companies who would be happy to accept this as a project. If anyone is interested in sponsoring this project, contact me and I'd be happy to give you a referral.

AndrewJo commented:
I apologize guys, but working at a startup takes up a HUGE chunk of my time. Working on a SQLite adapter was more of a pastime hobby for me for me to hone my JS skills and learns the innards of a well written ORM library. I will gladly take in PRs on my repo if it will help the project get past the finish line.

EhevuTov commented:
@AndrewJo, do you have time to resolve merge conflicts if there are any? I'm wondering if you would be more comfortable with someone else who has the time to maintain it and merging what you currently have in their repo.

mbret commented:
Hi @AndrewJo I plan to use sqlite for a medium size project which has a part based on sails. I wanted to know if your module was ready enough or will be in the next months ? Anyway thanks for everything you have done yet and wish you the best with your startup :)

dasilva333 committed:
tested and working copy, has been running stable for a few years now

dasilva333 committed:
removed all comments and unused code

AndrewJo committed:
Merge pull request #3 from dasilva333/master

tested and working copy, has been running stable for a few years now

kcappieg committed:
Upgrading sqlite3 to avoid windows install error

Old version of sqlite3 required dll on Windows

kcappieg committed:
Add waterline-errors to dependencies

kcappieg committed:
updating package.json for sails 1.x

kcappieg committed:
start refactor for waterline adapter api version 1

kcappieg committed:
add missing require

kcappieg committed:
attempt define method

I don't think it will work. Probably need to update utils for Waterline Adapter API

kcappieg committed:
add drop function

kcappieg committed:
async bug

kcappieg committed:
wrap drop call in spawn

kcappieg committed:
minor bug fixes, notes on other bugs

kcappieg committed:
updates to util for creating tables

automigrations *might* be ready

kcappieg committed:
bug fixes

kcappieg committed:
get describes in prep for create

kcappieg committed:
modify describe call

kcappieg committed:
merge from remote

kcappieg committed:
implement most of create

kcappieg committed:
bugfix async

kcappieg committed:
bug fixes and create method

kcappieg committed:
start update implementation

kcappieg committed:
bug fix update

kcappieg committed:
an attempt at find

kcappieg committed:
Basic find works

Don't try too much, including where conditions which are still untried

kcappieg committed:
refactor create into createEach

kcappieg committed:
bug fix find, update, createEach

find where criteria do not match current waterline api. must be migrated

kcappieg committed:
Begin refactor query.where

refactoring whole method for better logical structuring and new waterline where format

kcappieg committed:
Finish an implementation of where

Criteria tree evaluated recursively, includes all v1 operators, unsure about JSON column type

kcappieg committed:
Query.where bug fixes

All criteria pass smoke tests (No real testing though)

kcappieg committed:
destroy method

kcappieg committed:
bug fixes

kcappieg committed:
refactor for sequential writes

Sqlite is not made for concurrent writes - refactor to only spawn one write client

kcappieg committed:
Prepping for new tests

kcappieg committed:
Merge branch 'master' of github.com:kcappieg/sails-sqlite3

kcappieg committed:
Start testing w/ waterline tests

fix some bugs, discover others

kcappieg committed:
Bug fix createEach fetch

kcappieg committed:
Fix NOT IN error

Was coded as 'IN', added 'NOT'

kcappieg committed:
Passing all waterline tests

kcappieg committed:
Bug fixes for Associations

Adding associations interface tests

kcappieg committed:
Clean unused code, update readme

kcappieg committed:
readme updates

AndrewJo committed:
Merge pull request #4 from kcappieg/master

Waterline API v1 rewrite

kcappieg committed:
update repo

kcappieg committed:
0.1.1

kcappieg committed:
bug fix create method

kcappieg committed:
0.1.2

kcappieg committed:
update package versions

kcappieg committed:
0.1.3

Oxicode commented:
+1
